sound/soc/sof/sof-client-probes-ipc4.c

```diff
@@ -0,0 +1,289 @@
+	SOF_IPC4_PROBE_INJECTION_DMA = 1,
+	SOF_IPC4_PROBE_INJECTION_DMA_DETACH,
+	SOF_IPC4_PROBE_POINTS,
+	SOF_IPC4_PROBE_POINTS_DISCONNECT,
+};
+
+/* For getting probe dmas */
+struct sof_ipc4_probe_dma {
+	union sof_ipc4_connector_node_id node_id;
+	unsigned int dma_buffer_size;
+} __packed;
+
+struct sof_ipc4_probe_gtw_cfg {
+	union sof_ipc4_connector_node_id node_id;
+	u32 dma_buffer_size;
+} __packed;
+
+struct sof_ipc4_probe_cfg {
+	struct sof_ipc4_base_module_cfg base;
+	struct sof_ipc4_probe_gtw_cfg gtw_cfg;
+} __packed;
+
+enum sof_ipc4_probe_type {
+	SOF_IPC4_PROBE_TYPE_INPUT = 0,
+	SOF_IPC4_PROBE_TYPE_OUTPUT,
+	SOF_IPC4_PROBE_TYPE_INTERNAL
+};
+
+union sof_ipc4_probe_point_id {
+	u32 full_id;
+	struct {
+		u32 module_id:16;
+		u32 instance_id:8;
+		u32 type:2;
+		u32 index:6;
+	} fields;
+} __packed;
+
+struct sof_ipc4_probe_point {
+	union sof_ipc4_probe_point_id buffer_id;
+	uint32_t purpose;
+	uint32_t stream_tag;
+} __packed;
+
+#define INVALID_PIPELINE_ID      0xFF
+
+static struct sof_man4_module *sof_ipc4_probe_get_module_info(struct sof_client_dev *cdev)
+{
+	struct sof_probes_priv *priv = cdev->data;
+	struct snd_sof_dev *sdev = cdev->sdev;
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_ipc4_fw_module *fw_modules = ipc4_data->fw_modules;
+	static const guid_t probe_uuid =
+		GUID_INIT(0x7CAD0808, 0xAB10, 0xCD23,
+			  0xEF, 0x45, 0x12, 0xAB, 0x34, 0xCD, 0x56, 0xEF);
+	int i;
+
+	if (!priv->ipc_priv) {
+		if (!fw_modules) {
+			dev_err(sdev->dev, "no fw_module information\n");
+			return NULL;
+		}
+
+		/* set module info */
+		for (i = 0; i < ipc4_data->num_fw_modules; i++) {
+			if (guid_equal(&probe_uuid,
+				       &fw_modules[i].man4_module_entry.uuid)) {
+				priv->ipc_priv = &fw_modules[i].man4_module_entry;
+				break;
+			}
+		}
+	        if (!priv->ipc_priv)
+			dev_err(sdev->dev, "%s: no matching uuid found", __func__);
+	}
+
+	return (struct sof_man4_module *) priv->ipc_priv;
+}
+
+static int ipc4_probes_init(struct sof_client_dev *cdev, u32 stream_tag,
+			    size_t buffer_size)
+{
+	struct snd_sof_dev *sdev = cdev->sdev;
+	const struct sof_ipc_ops *iops = sdev->ipc->ops;
+	struct sof_man4_module *mentry =
+		sof_ipc4_probe_get_module_info(cdev);
+	struct sof_ipc4_msg msg;
+	struct sof_ipc4_probe_cfg cfg = {
+		.gtw_cfg = {
+			.node_id = {
+				.vindex = stream_tag - 1,
+				.dma_type = SOF_IPC4_DMA_HDA_HOST_INPUT,
+			},
+			.dma_buffer_size = buffer_size,
+		},
+	};
+	int ret;
+
+	if (!mentry)
+		return -EINVAL;
+
+	msg.primary = mentry->id;
+	msg.primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_INIT_INSTANCE);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+	msg.extension = SOF_IPC4_MOD_EXT_DST_MOD_INSTANCE(INVALID_PIPELINE_ID);
+	msg.extension |= SOF_IPC4_MOD_EXT_CORE_ID(0);
+
+	msg.data_size = sizeof(cfg);
+	msg.data_ptr = &cfg;
+
+	ret = iops->tx_msg(sdev, &msg, msg.data_size, NULL, 0, false);
```

Comment:
No, use the `sof_client_ipc_tx_message()`

Reply:
This is fine, in this place, but replacing  set_get_data() calls with sof_client_ipc_tx_message() is not that straight forwards. At least in theory the payload may not fit into a single message, and even worse I do not know how to check for that without accessing the sdev?  What about if I add the set_get_data into the sof client API?

Reply:
If the payload can not fit into a single payload than you can only use SOF_IPC4_MOD_LARGE_CONFIG_SET message which is completely different than the SOF_IPC4_MOD_INIT_INSTANCE.